fgpyo/io/io.py

```diff
@@ -0,0 +1,175 @@
+   ['flat file', '10']
+   >>> read_lines(path = path_compressed)
+   ['gzip file', '10']
+"""
+
+import gzip
+import io
+import os
+from pathlib import Path
+from typing import IO
+from typing import Any
+from typing import Generator
+from typing import Iterable
+from typing import Iterator
+from typing import Set
+from typing import TextIO
+from typing import Union
+from typing import cast
+
+COMPRESSED_FILE_EXTENSIONS: Set[str] = {".gz", ".bgz"}
+
+
+def path_is_readable(path: Path) -> None:
+    """Checks that file exists and returns True, else raises FileNotFoundError
+
+    Args:
+        path: a Path to check
+
+    Example:
+    _file_exists(path = Path("some_file.csv"))
+    """
+    assert path.exists(), f"Cannot read non-existent path: {path}"
+    assert not path.is_dir(), f"Cannot read path becasue it is a directory: {path}"
+    assert os.access(path, os.R_OK), f"Path exists but is not readable: {path}"
+
+
+def directory_exists(path: Path) -> None:
+    """Asserts that a path exist and is a directory
+
+    Args:
+        path: Path to check
+
+    Example:
+    _directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def path_is_writeable(path: Path, parent_must_exist: bool = True) -> None:
+    """Checks that path is writable and returns True, else raises an Exception
+    Args:
+         path: Path to check
+         parent_must_exist: True/False the parent directory must exist, the default is True
+    Example:
+    _writeable(path = Path("example.txt"))
+    """
+    assert path.exists(), f"Cannot write file because path is non-existent: {path}"
+    assert path.is_file(), f"Cannot write file because path is a directory: {path}"
+    assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+
+    if not parent_must_exist:
+        return None
+
+    parent: str = f"{path.parent.absolute()}"
+    assert Path(
+        parent
+    ).exists(), f"Cannot write file because parent diretory does not exist: {path}"
+    assert Path(
+        parent
+    ).is_dir(), f"Cannot write file because parent exists and is not a directory: {path}"
+    assert os.access(
+        parent, os.W_OK
+    ), f"Cannot write file because parent directory is not writeable: {path}"
+
+
+def to_reader(path: Path) -> Union[io.TextIOWrapper, TextIO, IO[Any]]:
+    """Opens a Path for reading with a specified mode or default 'rt'.
+
+    Args:
+        path: Path to read from
+
+    Example:
+    >>> reader = io.to_reader(path = Path("reader.txt"))
+    >>> reader.readlines()
+    >>> reader.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return io.TextIOWrapper(cast(IO[bytes], gzip.open(path, mode="rb")), encoding="utf-8")
+    else:
+        return path.open(mode="r")
+
+
+def to_writer(path: Path) -> Union[IO[Any], io.TextIOWrapper]:
```

Comment:
Same comment here about mode.
Also, perhaps this extends to both `open` functions - should they return a class that implements `ContextManager`?  Or do all types in the union(s) work as context managers, such that you could write `with io.open_for_text_writing(path) as fg`?